include/seqan3/range/view/take_until.hpp

```diff
@@ -170,6 +167,135 @@ class view_take_until : public ranges::view_interface<view_take_until<urng_t, fu
+        constexpr iterator_type_consume_input & operator=(iterator_type_consume_input && rhs) = default; //!< Defaulted.
+        ~iterator_type_consume_input() = default;
+
+        //!\brief Constructor that delegates to the CRTP layer.
+        iterator_type_consume_input(base_base_t const & it) :
+            base_t{it}
+        {}
+
+        //!\brief Constructor that delegates to the CRTP layer and initialises the callable.
+        iterator_type_consume_input(base_base_t it, fun_ref_t _fun, sentinel_type sen) :
+            base_t{std::move(it)}, fun{_fun}, stored_end{std::move(sen)}
+        {}
+        //!\}
+
+        /*!\name Associated types
+         * \brief All are derived from the base_base_t.
+         * \{
+         */
+        using difference_type       = typename std::iterator_traits<base_base_t>::difference_type;  //!< From base.
+        using value_type            = typename std::iterator_traits<base_base_t>::value_type;       //!< From base.
+        using reference             = typename std::iterator_traits<base_base_t>::reference;        //!< From base.
+        using pointer               = typename std::iterator_traits<base_base_t>::pointer;          //!< From base.
+        using iterator_category     = std::input_iterator_tag;                                      //!< Always input.
+        //!\}
+
+        /*!\name Arithmetic operators
+         * \brief seqan3::detail::inherited_iterator_base operators are used unless specialised here.
+         * \{
+         */
+        //!\brief Override pre-increment to implement consuming behaviour.
+        iterator_type_consume_input & operator++() noexcept(!or_throw)
+        {
+            ++static_cast<base_base_t &>(*this);
+
+            if ((static_cast<base_base_t &>(*this) != stored_end) && fun(**this))
+            {
+                at_end_gracefully = true;
+                ++(*this); // call this function recursively
+            }
+
+            return *this;
+        }
+
+        //!\brief Post-increment implemented via pre-increment.
+        iterator_type_consume_input operator++(int)
+        {
+            iterator_type_consume_input cpy{*this};
+            ++(*this);
+            return cpy;
+        }
+        //!\}
+
+        /*!\name Comparison operators
+         * \brief We define comparison against self and against the sentinel.
+         * \{
+         */
+        //!\brief Return the saved at_end state.
+        bool operator==(sentinel_type const & rhs) const noexcept(!or_throw)
+        {
+            if (at_end_gracefully)
+                return true;
+
+            if (static_cast<base_base_t const &>(*this) == rhs)
+            {
+                if constexpr (or_throw)
+                    throw unexpected_end_of_input{"Reached end of input before functor evaluated to true."};
+                else
+                    return true;
+            }
+
+            return fun(**this);
```

Comment:
Is this ever true? because the moviing operations (++ etc.) always check `fun` and will set `at_end_gracefully`...

Reply:
not resolved?